libfec/src/viterbi27.c

```diff
@@ -0,0 +1,200 @@
+/* K=7 r=1/2 Viterbi decoder in portable C
+ * Copyright Feb 2004, Phil Karn, KA9Q
+ * May be used under the terms of the GNU Lesser General Public License (LGPL)
+ */
+#include <stdio.h>
+#include <stdlib.h>
+#include <memory.h>
+#include <limits.h>
+#include "fec.h"
+
+
+typedef union {
+  unsigned int w[64];
+} metric_t;
+
+typedef union {
+  unsigned long w[2];
+} decision_t;
+
+static union branchtab27 {
+  unsigned char c[32];
+} Branchtab27[2] __attribute__ ((aligned(16)));
+
+static int Init = 0;
+
+/* State info for instance of Viterbi decoder
+ */
+struct v27 {
+  metric_t metrics1; /* path metric buffer 1 */
+  metric_t metrics2; /* path metric buffer 2 */
+  decision_t *dp;          /* Pointer to current decision */
+  metric_t *old_metrics,*new_metrics; /* Pointers to path metrics, swapped on every bit */
+  decision_t *decisions;   /* Beginning of decisions for block */
+};
+
+/* Initialize Viterbi decoder for start of new frame */
+int init_viterbi27(void *p,int starting_state){
+  struct v27 *vp = p;
+  int i;
+
+  if(p == NULL)
+    return -1;
+  for(i=0;i<64;i++)
+    vp->metrics1.w[i] = 63;
+
+  vp->old_metrics = &vp->metrics1;
+  vp->new_metrics = &vp->metrics2;
+  vp->dp = vp->decisions;
+  vp->old_metrics->w[starting_state & 63] = 0; /* Bias known start state */
+  return 0;
+}
+
+void set_viterbi27_polynomial(int polys[2]){
+  int state;
+
+  for(state=0;state < 32;state++){
+    Branchtab27[0].c[state] = (polys[0] < 0) ^ parity((2*state) & abs(polys[0])) ? 255 : 0;
+    Branchtab27[1].c[state] = (polys[1] < 0) ^ parity((2*state) & abs(polys[1])) ? 255 : 0;
+  }
+  Init++;
+}
+
+/* Create a new instance of a Viterbi decoder */
+void *create_viterbi27(int len){
+  struct v27 *vp;
+
+  if(!Init){
+    int polys[2] = { V27POLYA, V27POLYB };
+    set_viterbi27_polynomial(polys);
+  }
+  if((vp = malloc(sizeof(struct v27))) == NULL)
```

Comment:
I'm inclined somewhat against using cmake flags for this kind of thing.
Maybe the length could stay flexible but the firmware could reserve memory for the decision table and pass in a pointer to it.

As to what the length should actually be: See if you can figure out a way to daisy-chain processing of blocks without losing data bits.  So we could have the FEC decode block arbitrarily short (shorter just means slightly lower error-correction performance).  3 seconds' worth of data is definitely too much, that would mean 250*3*sizeof(decision_t) = 6 kilobytes which we can't afford to spare.  Maybe also see if it's straightforward to use u16s for the decision table rather than u32s.